src/MongoDB.Driver.Core/Core/WireProtocol/Messages/Encoders/BinaryEncoders/CompressedMessageBinaryEncoder.cs

```diff
@@ -0,0 +1,226 @@
+        private const int MessageHeaderLength = 16;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CompressedMessageBinaryEncoder" /> class.
+        /// </summary>
+        /// <param name="stream">The stream.</param>
+        /// <param name="compressorSource">The compressor source.</param>
+        /// <param name="encoderSettings">The encoder settings.</param>
+        public CompressedMessageBinaryEncoder(
+            Stream stream,
+            ICompressorSource compressorSource,
+            MessageEncoderSettings encoderSettings)
+            : base(stream, encoderSettings)
+        {
+            _compressorSource = Ensure.IsNotNull(compressorSource, nameof(compressorSource));
+            _encoderSettings = encoderSettings; // can be null
+        }
+
+        /// <summary>
+        /// Gets a flag whether the message compressed or not.
+        /// </summary>
+        /// <returns>A compressed flag.</returns>
+        public bool IsCompressed()
+        {
+            var stream = CreateBinaryReader().BsonStream;
+
+            var position = stream.Position;
+
+            // Skip header to opcode
+            stream.ReadInt32();
+            stream.ReadInt32();
+            stream.ReadInt32();
+            var isCompressed = stream.ReadInt32() == (int)Opcode.Compressed;
+
+            stream.Position = position;
+
+            return isCompressed;
+        }
+
+        /// <summary>
+        /// Reads the message.
+        /// </summary>
+        /// <returns>A message.</returns>
+        public CompressedMessage ReadMessage()
+        {
+            var stream = CreateBinaryReader().BsonStream;
+            var compressedMessageInfo = ReadMessage(stream);
+
+            var compressor = _compressorSource.Get(compressedMessageInfo.CompressorType);
+            var uncompressedBytes = compressor.Decompress(compressedMessageInfo.CompressedBytes);
+
+            using (var chunkSource = new OutputBufferChunkSource(BsonChunkPool.Default))
+            using (var buffer = new MultiChunkBuffer(chunkSource))
+            using (var originalMessageStream = CreateStreamWithOriginalHeaders(buffer, compressedMessageInfo))
+            {
+                originalMessageStream.WriteBytes(uncompressedBytes.Array, 0, uncompressedBytes.Count);
+
+                originalMessageStream.Position = 0;
+                buffer.MakeReadOnly();
+
+                var encoder = GetEncoderByOpcode(compressedMessageInfo.OriginalOpCode, originalMessageStream);
+                var originalMessage = encoder.ReadMessage();
+
+                originalMessageStream.Position = 0;
+                using (var binaryReader = new BinaryReader(originalMessageStream))
+                {
+                    return new CompressedMessage(
+                        originalMessage,
+                        new ArraySegment<byte>(binaryReader.ReadBytes((int)originalMessageStream.Length), 0, (int)originalMessageStream.Length),
```

Comment:
Probably, we will not have problems with converting `long` to `int` here?

Reply:
In theory we could overflow.

Messages exchanged with the server are limited to 48MB. If Compression fails to compress they could be larger, but probably not large enough to overflow.

If you want to add code to check for overflow that would be good.